File: with-streaming/streaming.ts
/**
 * with-streaming/streaming.ts — 스트리밍 예제 (TypeScript)
 *
 * 두 가지 패턴을 보여준다:
 * 1. 실시간 스트리밍: StreamEvent에서 텍스트 청크를 추출해 즉시 출력
 * 2. 블로킹 vs 스트리밍 비교: 첫 출력까지의 시간(TTFB) 측정
 */

import { query } from "@anthropic-ai/claude-agent-sdk";
import { printResult } from "../utils/utils.ts";

type AnyMessage = Record<string, unknown>;

async function streamResponse(): Promise<void> {
  /**
   * 패턴 1: StreamEvent로 텍스트 실시간 출력.
   *
   * includePartialMessages: true 설정 시 StreamEvent가 yield된다.
   * event.type === "content_block_delta" 이고
   * event.delta.type === "text_delta" 인 경우 텍스트 청크가 담겨 있다.
   *
   * 출력 순서:
   * - StreamEvent: 텍스트가 생성될 때마다 도착 (부분 텍스트)
   * - AssistantMessage: 완성된 전체 메시지 (스트리밍 완료 후)
   * - ResultMessage: 완료 신호 + 비용 정보
   */
  console.log("=".repeat(50));
  console.log("패턴 1: 실시간 스트리밍");
  console.log("=".repeat(50));
  console.log("[스트리밍 시작 →]\n");

  for await (const message of query({
    prompt:
      "Python의 asyncio 이벤트 루프가 어떻게 동작하는지 " +
      "단계별로 자세히 설명해주세요.",
    options: {
      // 이 옵션을 true로 설정해야 StreamEvent가 yield된다
      includePartialMessages: true,
    },
  })) {
    const msg = message as AnyMessage;

    if (msg.type === "stream_event") {
      const event = msg.event as AnyMessage;
      // content_block_delta 이벤트에 실제 텍스트 청크가 담긴다
      if (event.type === "content_block_delta") {
        const delta = event.delta as AnyMessage;
        if (delta.type === "text_delta") {
          // flush 없이도 Node.js stdout은 즉시 출력된다
          process.stdout.write(delta.text as string);
        }
        // thinking_delta: Extended Thinking 활성화 시 사고 과정 스트리밍
      }
    } else if (msg.type === "assistant") {
      // StreamEvent가 모두 도착한 뒤 완성된 메시지가 온다
      process.stdout.write("\n\n");
    } else if (msg.type === "result") {
      printResult(msg);
    }
  }
}

async function compareBlockingVsStreaming(): Promise<void> {
  /**
   * 패턴 2: 블로킹 방식과 스트리밍 방식의 체감 속도 비교.
   * 동일한 프롬프트로 두 방식을 실행하고 TTFB를 측정한다.
   */
  const prompt = "Python 제너레이터의 장점을 세 가지만 설명해주세요.";

  // --- 블로킹 방식 ---
  console.log("\n" + "=".repeat(50));
  console.log("패턴 2a: 블로킹 방식 (스트리밍 없음)");
  console.log("=".repeat(50));

  let start = Date.now();
  let firstOutputTime: number | null = null;

  for await (const message of query({
    prompt,
    options: { includePartialMessages: false },
  })) {
    const msg = message as AnyMessage;

    if (msg.type === "assistant") {
      const elapsed = (Date.now() - start) / 1000;
      if (firstOutputTime === null) {
        firstOutputTime = elapsed;
        console.log(`[첫 출력까지: ${elapsed.toFixed(2)}초]`);
      }
      const content = (msg.message as { content: AnyMessage[] }).content;
      for (const block of content) {
        if (block.type === "text") console.log(block.text as string);
      }
    }
  }
  console.log(`[총 소요 시간: ${((Date.now() - start) / 1000).toFixed(2)}초]`);

  // --- 스트리밍 방식 ---
  console.log("\n" + "=".repeat(50));
  console.log("패턴 2b: 스트리밍 방식");
  console.log("=".repeat(50));

  start = Date.now();
  firstOutputTime = null;

  for await (const message of query({
    prompt,
    options: { includePartialMessages: true },
  })) {
    const msg = message as AnyMessage;

    if (msg.type === "stream_event") {
      const event = msg.event as AnyMessage;
      if (
        event.type === "content_block_delta" &&
        (event.delta as AnyMessage)?.type === "text_delta"
      ) {
        const elapsed = (Date.now() - start) / 1000;
        if (firstOutputTime === null) {
          firstOutputTime = elapsed;
          console.log(`[첫 출력까지: ${elapsed.toFixed(2)}초]`);
        }
        process.stdout.write((event.delta as AnyMessage).text as string);
      }
    } else if (msg.type === "assistant") {
      process.stdout.write("\n");
    }
  }

  console.log(`\n[총 소요 시간: ${((Date.now() - start) / 1000).toFixed(2)}초]`);
  console.log("\n[결론: 총 시간은 비슷하지만 스트리밍 방식이 첫 출력이 훨씬 빠름]");
}

async function main(): Promise<void> {
  await streamResponse();
  await compareBlockingVsStreaming();
}

main().catch(console.error);

File: with-streaming/streaming.py
"""
with-streaming/streaming.py — 스트리밍 예제 (Python)

두 가지 패턴을 보여준다:
1. 실시간 스트리밍: StreamEvent에서 텍스트 청크를 추출해 즉시 출력
2. 블로킹 vs 스트리밍 비교: 두 방식의 체감 응답 속도 차이
"""

import asyncio
import sys
import time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent))

from claude_agent_sdk import (
    AssistantMessage,
    ClaudeAgentOptions,
    ResultMessage,
    StreamEvent,
    query,
)
from claude_agent_sdk.types import TextBlock

from utils.utils import print_result


async def stream_response() -> None:
    """
    패턴 1: StreamEvent로 텍스트 실시간 출력.

    include_partial_messages=True 설정 시 StreamEvent가 yield된다.
    StreamEvent.event 딕셔너리에서 text_delta를 추출해 즉시 출력한다.

    출력 구조:
    - StreamEvent: 텍스트가 생성될 때마다 도착 (부분 텍스트)
    - AssistantMessage: 완성된 전체 메시지 (스트리밍 완료 후)
    - ResultMessage: 완료 신호 + 비용 정보
    """
    print("=" * 50)
    print("패턴 1: 실시간 스트리밍")
    print("=" * 50)
    print("[스트리밍 시작 →]\n")

    options = ClaudeAgentOptions(
        # 이 옵션을 True로 설정해야 StreamEvent가 yield된다
        include_partial_messages=True,
    )

    async for message in query(
        prompt=(
            "Python의 asyncio 이벤트 루프가 어떻게 동작하는지 "
            "단계별로 자세히 설명해주세요."
        ),
        options=options,
    ):
        if isinstance(message, StreamEvent):
            event = message.event
            # content_block_delta 이벤트에 실제 텍스트 청크가 담긴다
            if event.get("type") == "content_block_delta":
                delta = event.get("delta", {})
                if delta.get("type") == "text_delta":
                    # flush=True: 버퍼 없이 즉시 출력
                    print(delta["text"], end="", flush=True)
                elif delta.get("type") == "thinking_delta":
                    # Extended Thinking 활성화 시 사고 과정도 스트리밍됨
                    pass

        elif isinstance(message, AssistantMessage):
            # StreamEvent가 모두 도착한 뒤 완성된 메시지가 온다
            # 이미 스트리밍으로 출력했으므로 여기서는 줄바꿈만
            print("\n")

        elif isinstance(message, ResultMessage):
            print_result(message)


async def compare_blocking_vs_streaming() -> None:
    """
    패턴 2: 블로킹 방식과 스트리밍 방식의 체감 속도 비교.

    동일한 프롬프트로 두 방식을 실행하고 첫 출력까지의 시간(TTFB)을 측정한다.
    실제 API 처리 시간은 같지만 사용자 체감 속도는 크게 다르다.
    """
    prompt = "Python 제너레이터의 장점을 세 가지만 설명해주세요."

    # --- 블로킹 방식 ---
    print("\n" + "=" * 50)
    print("패턴 2a: 블로킹 방식 (스트리밍 없음)")
    print("=" * 50)

    start = time.time()
    first_output_time: float | None = None

    async for message in query(
        prompt=prompt,
        options=ClaudeAgentOptions(include_partial_messages=False),
    ):
        if isinstance(message, AssistantMessage):
            elapsed = time.time() - start
            if first_output_time is None:
                first_output_time = elapsed
                print(f"[첫 출력까지: {elapsed:.2f}초]")
            for block in message.content:
                if isinstance(block, TextBlock):
                    print(block.text)

    print(f"[총 소요 시간: {time.time() - start:.2f}초]")

    # --- 스트리밍 방식 ---
    print("\n" + "=" * 50)
    print("패턴 2b: 스트리밍 방식")
    print("=" * 50)

    start = time.time()
    first_output_time = None

    async for message in query(
        prompt=prompt,
        options=ClaudeAgentOptions(include_partial_messages=True),
    ):
        if isinstance(message, StreamEvent):
            event = message.event
            if (
                event.get("type") == "content_block_delta"
                and event.get("delta", {}).get("type") == "text_delta"
            ):
                elapsed = time.time() - start
                if first_output_time is None:
                    first_output_time = elapsed
                    print(f"[첫 출력까지: {elapsed:.2f}초]")
                print(event["delta"]["text"], end="", flush=True)

        elif isinstance(message, AssistantMessage):
            print()

    print(f"\n[총 소요 시간: {time.time() - start:.2f}초]")
    print("\n[결론: 총 시간은 비슷하지만 스트리밍 방식이 첫 출력이 훨씬 빠름]")


async def main() -> None:
    await stream_response()
    await compare_blocking_vs_streaming()


if __name__ == "__main__":
    asyncio.run(main())
